cpp/include/raft/core/bitset.cuh

```diff
@@ -0,0 +1,308 @@
+                      (new_size - old_size) * sizeof(bitset_t),
+                      resource::get_cuda_stream(res));
+    }
+  }
+
+  /**
+   * @brief Test a list of indices in a bitset.
+   *
+   * @tparam output_t Output type of the test. Default is bool.
+   * @param res RAFT resources
+   * @param queries List of indices to test
+   * @param output List of outputs
+   */
+  template <typename output_t = bool>
+  void test(const raft::resources& res,
+            raft::device_vector_view<const index_t, index_t> queries,
+            raft::device_vector_view<output_t, index_t> output) const
+  {
+    RAFT_EXPECTS(output.extent(0) == queries.extent(0), "Output and queries must be same size");
+    auto bitset_view = view();
+    raft::linalg::map(
+      res,
+      output,
+      [bitset_view] __device__(index_t query) { return output_t(bitset_view.test(query)); },
+      queries);
+  }
+  /**
+   * @brief Set a list of indices in a bitset to set_value.
+   *
+   * @param res RAFT resources
+   * @param mask_index indices to remove from the bitset
+   * @param set_value Value to set the bits to (true or false)
+   */
+  void set(const raft::resources& res,
+           raft::device_vector_view<const index_t, index_t> mask_index,
+           bool set_value = false)
+  {
+    auto* bitset_ptr = this->data_handle();
+    thrust::for_each_n(resource::get_thrust_policy(res),
+                       mask_index.data_handle(),
+                       mask_index.extent(0),
+                       [bitset_ptr, set_value] __device__(const index_t sample_index) {
+                         const index_t bit_element = sample_index / bitset_element_size;
+                         const index_t bit_index   = sample_index % bitset_element_size;
+                         const bitset_t bitmask    = bitset_t{1} << bit_index;
+                         if (set_value) {
+                           atomicOr(bitset_ptr + bit_element, bitmask);
+                         } else {
+                           const bitset_t bitmask2 = ~bitmask;
+                           atomicAnd(bitset_ptr + bit_element, bitmask2);
+                         }
+                       });
+  }
+  /**
+   * @brief Flip all the bits in a bitset.
+   *
+   * @param res RAFT resources
+   */
+  void flip(const raft::resources& res)
+  {
+    auto bitset_span = this->to_mdspan();
+    raft::linalg::map(
+      res,
+      bitset_span,
+      [] __device__(bitset_t element) { return bitset_t(~element); },
+      raft::make_const_mdspan(bitset_span));
+  }
+  /**
+   * @brief Reset the bits in a bitset.
+   *
+   * @param res RAFT resources
+   */
+  void reset(const raft::resources& res)
+  {
+    cudaMemsetAsync(bitset_.data(),
+                    default_value_ ? 0xff : 0x00,
+                    n_elements() * sizeof(bitset_t),
+                    resource::get_cuda_stream(res));
+  }
+
+ private:
+  raft::device_uvector<bitset_t> bitset_;
```

Comment:
What type is this? I don't recognize it

Reply:
It's a wrapper around `rmm::device_uvector`